renderdoc/driver/vulkan/vk_pixelhistory.cpp

```diff
@@ -163,9 +348,61 @@ struct VulkanPixelHistoryCallback : public VulkanDrawcallCallback
     m_pDriver->GetRenderState().EndRenderPass(cmd);
     // Get pre-modification values
     size_t storeOffset = m_EventIndices.size() * sizeof(EventInfo);
-    CopyPixel(eid, cmd, storeOffset);
+    VkImage depthImage = VK_NULL_HANDLE;
+    VkFormat depthFormat = VK_FORMAT_UNDEFINED;
+    const DrawcallDescription *draw = m_pDriver->GetDrawcall(eid);
+    if(draw && draw->depthOut != ResourceId())
+    {
+      ResourceId resId = m_pDriver->GetResourceManager()->GetLiveID(draw->depthOut);
+      depthImage = m_pDriver->GetResourceManager()->GetCurrentHandle<VkImage>(resId);
+      const VulkanCreationInfo::Image &imginfo = m_pDriver->GetDebugManager()->GetImageInfo(resId);
+      depthFormat = imginfo.format;
+    }
+
+    CopyPixel(m_Image, m_Format, depthImage, depthFormat, cmd, storeOffset);
 
     // TODO: add all logic for draw calls.
+    VulkanRenderState &pipestate = m_pDriver->GetRenderState();
+    ResourceId prevState = pipestate.graphics.pipeline;
+    ResourceId prevRenderpass = pipestate.renderPass;
+    ResourceId prevFramebuffer = pipestate.framebuffer;
+    uint32_t prevSubpass = pipestate.subpass;
+
+    {
+      auto pipeIt = m_PipelineCache.find(pipestate.graphics.pipeline);
+      PipelineReplacements replacements;
+
+      if(pipeIt != m_PipelineCache.end())
+        replacements = pipeIt->second;
+      else
+      {
+        // TODO: Create other pipeline replacements
+        replacements.occlusion = CreatePipelineForOcclusion(pipestate.graphics.pipeline);
+
+        m_PipelineCache.insert(std::make_pair(pipestate.graphics.pipeline, replacements));
+      }
+
+      const VulkanCreationInfo::Pipeline &p =
+          pipestate.m_CreationInfo->m_Pipeline[pipestate.graphics.pipeline];
+      if(p.dynamicStates[VkDynamicViewport])
+        for(uint32_t i = 0; i < pipestate.views.size(); i++)
+          UpdateScissor(pipestate.views[i], pipestate.scissors[i]);
+
+      pipestate.framebuffer = GetResID(m_OffscreenFB);
+      pipestate.renderPass = GetResID(m_RenderPass);
+      pipestate.subpass = 0;
+      pipestate.graphics.pipeline = GetResID(replacements.occlusion);
+      ReplayDraw(cmd, (uint32_t)m_OcclusionQueries.size(), eid, true);
```

Comment:
We have to be very careful when replaying draws multiple times, because draws can have side-effects more than just their rasterized outputs to the framebuffer. Even rebinding an offscreen framebuffer and renderpass, any storage images and buffers will be written to by the draw twice which might lead to a diverging replay.

On D3D11 this isn't really a problem because it's easy to unbind all UAVs and let writes become no-ops, and for UAV output bindings anyway we skip the occlusion query. On Vulkan though it's hard to unbind all storage images and we don't have the "stores to unbound resources are dropped" behaviour to fall back on anyway.

One alternative we have on Vulkan is to patch all the shaders' SPIR-V to remove all stores. We've already replaced the fragment shader with a fixed color, so this would mean patching the other shaders in the pipeline to remove any `OpStore` type operation to a buffer/image.

Reply:
That makes sense, thanks! Added a TODO to address this.